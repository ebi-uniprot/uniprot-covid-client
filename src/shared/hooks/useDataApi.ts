import { useEffect, useReducer } from 'react';
import axios, { AxiosResponse, AxiosError } from 'axios';

import fetchData from '../utils/fetchData';

type State = {
  loading: boolean;
  data?: AxiosResponse['data'];
  status?: AxiosResponse['status'];
  statusText?: AxiosResponse['statusText'];
  headers?: AxiosResponse['headers'];
  error?: AxiosError;
  redirectedTo?: string;
};

enum ActionType {
  INIT = 'INIT',
  SUCCESS = 'SUCCESS',
  ERROR = 'ERROR',
}

type Action =
  | { type: ActionType.INIT }
  | { type: ActionType.SUCCESS; response?: AxiosResponse; originalURL?: string }
  | { type: ActionType.ERROR; error: AxiosError };

// eslint-disable-next-line consistent-return
const reducer = (state: State, action: Action): State => {
  // eslint-disable-next-line default-case
  switch (action.type) {
    case ActionType.INIT:
      return {
        loading: true,
      };
    case ActionType.SUCCESS:
      // eslint-disable-next-line no-case-declarations
      const newState: State = {
        loading: false,
        data: action.response && action.response.data,
        status: action.response && action.response.status,
        statusText: action.response && action.response.statusText,
        headers: action.response && action.response.headers,
      };
      if (
        action.response &&
        action.response.request.responseURL &&
        action.response.request.responseURL !== action.originalURL
      ) {
        newState.redirectedTo = action.response.request.responseURL;
      } else if (
        // Issue with casing in axios-mock-adapter?
        action.response &&
        action.response.request.responseUrl &&
        action.response.request.responseUrl !== action.originalURL
      ) {
        newState.redirectedTo = action.response.request.responseUrl;
      }
      return newState;
    case ActionType.ERROR:
      return {
        loading: false,
        status: action.error.response && action.error.response.status,
        statusText: action.error.response && action.error.response.statusText,
        headers: action.error.response && action.error.response.headers,
        error: action.error,
      };
  }
};

const useDataApi = (url?: string): State => {
  const [state, dispatch] = useReducer(reducer, { loading: !!url });

  useEffect(() => {
    // we don't require a URL, we just don't need data anymore
    // assume succes with no data
    if (!url) {
      dispatch({ type: ActionType.SUCCESS });
      return;
    }

    dispatch({ type: ActionType.INIT });

    // variables to handle cancellation
    const source = axios.CancelToken.source();

    // actual request
    fetchData(url, undefined, source.token).then(
      // handle ok
      (response: AxiosResponse) =>
        dispatch({ type: ActionType.SUCCESS, response, originalURL: url }),
      // catch error
      (error: AxiosError) => {
        if (axios.isCancel(error)) return;
        dispatch({ type: ActionType.ERROR, error });
      }
    );

    // handle unmounting of the hook
    // eslint-disable-next-line consistent-return
    return () => source.cancel();
  }, [url]);

  return state;
};

export default useDataApi;
