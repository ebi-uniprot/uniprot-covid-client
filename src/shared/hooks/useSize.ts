import { MutableRefObject, useCallback, useEffect } from 'react';
import useSafeState from './useSafeState';

type ContainerElement = HTMLElement | SVGElement | SVGSVGElement;

/**
 * given a reference to an actual HTML element, gives us access to its size
 * even through resizes and zoom
 * @param {MutableRefObject<HTMLElement>} ref
 * @returns {Array} tuple - return tuple
 * @returns {DOMRect | null} tuple[0]
 *          DOMRect object with the sizes of the passed ref, or null if no ref
 * @returns {() => void} tuple[1]
 *          Manual refresh function, to call if the size needs to be recomputed
 *          for whatever reason (e.g. content changed)
 */
function useSize<E extends ContainerElement = ContainerElement>(
  ref: MutableRefObject<E | undefined | null>
): [DOMRect | null, () => void] {
  const [rect, setRect] = useSafeState<DOMRect | null>(null);

  const onResize = useCallback(() => {
    if (!ref.current) {
      return;
    }
    setRect(ref.current.getBoundingClientRect());
  }, [ref, setRect]);

  // This effect will run after EVERY render
  useEffect(() => {
    // ask for size value until we finally get one
    if (!rect) {
      onResize();
    }
  }); // No dependency array, it's on purpose!

  useEffect(() => {
    onResize(); // first time
    window.addEventListener('resize', onResize);
    // eslint-disable-next-line consistent-return
    return () => window.removeEventListener('resize', onResize);
  }, [ref, onResize]);

  return [rect, onResize];
}

export default useSize;
