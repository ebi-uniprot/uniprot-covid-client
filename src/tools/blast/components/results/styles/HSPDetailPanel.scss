.hsp-detail-panel {
  height: auto;
  max-height: 50%;

  &__placeholder {
    height: 80px;
  }

  &__header {
    display: flex;
    justify-content: space-between;
    padding-top: 0.5rem;
    padding-left: 0.5rem;
    padding-right: 0.5rem;
  }

  &__body {
    overflow-y: auto;
    max-height: 40vh;
  }

  &__visualisation {
    display: grid;
    grid-template-columns: 1fr 9fr;
    margin-bottom: 1em;
    align-items: center;

    &--msa {
      grid-area: msa;
    }

    &--track {
      grid-area: track;
    }

    &__wrapped-row {
      grid-template-areas:
        'msa-label msa'
        'track-label track';

      protvista-navigation {
        .zoom-polygon,
        .brush,
        .start-label,
        .end-label {
          display: none;
        }
        height: 1.25rem;
        padding-bottom: 0.25rem;
      }
    }
  }

  .hsp-label {
    text-align: right;
    padding-right: 1rem;
    font-weight: 800;
    &:first-letter {
      text-transform: uppercase;
    }

    &--msa {
      grid-area: msa-label;
    }

    &--track {
      grid-area: track-label;
    }
  }
}
