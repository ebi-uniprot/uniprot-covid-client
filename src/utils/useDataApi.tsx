/* eslint-disable @typescript-eslint/no-explicit-any */
import { useState, useEffect } from 'react';
import fetchData from './fetchData';

const useDataApi = (url: string) => {
  const [data, setData] = useState<any>({});

  useEffect(() => {
    const getData = async () => {
      try {
        const result = await fetchData(url);
        setData(result.data);
      } catch (error) {
        if (error.response.status !== '404') console.error(error);
        setData(null);
      }
    };
    getData();
  }, [url]);

  return data;
};

export default useDataApi;
