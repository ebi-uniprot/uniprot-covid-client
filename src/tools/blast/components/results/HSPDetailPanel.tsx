/* eslint-disable no-param-reassign */
/* eslint-disable camelcase */
import React, { FC } from 'react';
import { Loader, CloseIcon } from 'franklin-sites';
import SlidingPanel from '../../../../shared/components/layouts/SlidingPanel';
import { BlastHsp } from '../../types/blastResults';
import useDataApi from '../../../../shared/hooks/useDataApi';
import { UniProtkbAPIModel } from '../../../../uniprotkb/adapters/uniProtkbConverter';
import { getAccessionsURL } from '../../../../uniprotkb/config/apiUrls';
import ErrorHandler from '../../../../shared/components/error-pages/ErrorHandler';
import './styles/HSPDetailPanel.scss';
import MSAWrapper, { MSAInput, View } from '../../../components/MSAWrapper';

type UniProtkbAccessionsAPI = {
  results: UniProtkbAPIModel[];
};

export type HSPDetailPanelProps = {
  hsp: BlastHsp;
  hitAccession: string;
  extra?: UniProtkbAPIModel;
  onClose: () => void;
  hitLength: number;
  queryLength: number;
};

export const convertHSPtoMSAInputs = (
  hsp: BlastHsp,
  queryLength: number,
  hitLength: number,
  hitAccession: string,
  extra?: UniProtkbAPIModel
) => {
  const {
    hsp_query_from,
    hsp_query_to,
    hsp_qseq,
    hsp_hit_from,
    hsp_hit_to,
    hsp_hseq,
  } = hsp;

  return [
    {
      name: 'Query',
      sequence: hsp_qseq,
      from: hsp_query_from,
      to: hsp_query_to,
      length: queryLength,
    },
    {
      name: 'Match',
      sequence: hsp_hseq,
      from: hsp_hit_from,
      to: hsp_hit_to,
      length: hitLength,
      accession: hitAccession,
      features: extra?.features,
    },
  ];
};

const HSPDetailPanel: FC<HSPDetailPanelProps> = ({
  hsp,
  hitAccession,
  onClose,
  hitLength,
  queryLength,
  extra,
}) => {
  const { hsp_align_len } = hsp;

  // const totalLength = getFullAlignmentLength(hsp, queryLength, hitLength);

  // Reset view when different hit is being viewed
  // useEffect(() => {
  //   setActiveView(View.overview);
  //   setAnnotation(undefined);
  //   setHighlightProperty(undefined);
  // }, [hitAccession]);

  const { loading, data, status, error } = useDataApi<UniProtkbAccessionsAPI>(
    getAccessionsURL([hitAccession], { facets: [] })
  );

  const recommendedName =
    data?.results?.[0]?.proteinDescription?.recommendedName?.fullName.value;
  const organism = data?.results?.[0]?.organism?.scientificName;

  const title = [hitAccession, recommendedName, organism]
    .filter(Boolean)
    .join(' Â· ');

  const alignment: MSAInput[] = convertHSPtoMSAInputs(
    hsp,
    queryLength,
    hitLength,
    hitAccession,
    extra
  );

  if (error) {
    return <ErrorHandler status={status} />;
  }

  if (loading) {
    return <Loader />;
  }

  return (
    <SlidingPanel position="bottom" className="hsp-detail-panel">
      <div className="hsp-detail-panel__header">
        <h4>{title}</h4>
        <button type="button" onClick={onClose}>
          <CloseIcon width="16" height="16" />
        </button>
      </div>
      <div className="hsp-detail-panel__body">
        <MSAWrapper
          alignmentLength={hsp_align_len}
          // totalLength={totalLength}
          alignment={alignment}
          defaultView={View.overview}
        />
      </div>
    </SlidingPanel>
  );
};

export default HSPDetailPanel;
