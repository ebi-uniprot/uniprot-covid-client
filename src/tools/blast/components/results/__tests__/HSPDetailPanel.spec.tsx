import React from 'react';
import { fireEvent } from '@testing-library/react';
import blastResultsMockData from '../../../../__mocks__/server-jobs/example-truncated.json';
import HSPDetailPanel, { transformFeaturesPositions } from '../HSPDetailPanel';
import modelData from '../../../../../uniprotkb/__mocks__/entryModelData.json';
import renderWithRouter from '../../../../../shared/__test-helpers__/RenderWithRouter';

jest.mock('../../../../../shared/hooks/useDataApi', () => jest.fn());
jest.mock('../../../../../shared/hooks/useSize', () => jest.fn());
import useDataApi from '../../../../../shared/hooks/useDataApi';
import useSize from '../../../../../shared/hooks/useSize';

const dataMock = {
  loading: false,
  data: modelData,
};
useDataApi.mockImplementation(() => dataMock);
useSize.mockImplementation(() => [{ width: 1000 }]);

describe('HSPDetailPanel', () => {
  let rendered;
  const onClose = jest.fn();
  const hit = blastResultsMockData.hits[0];
  const hsp = hit.hit_hsps[0];

  beforeEach(async () => {
    rendered = renderWithRouter(
      <HSPDetailPanel
        hsp={hsp}
        hitAccession={hit.hit_acc}
        onClose={onClose}
        hitLength={hit.hit_len}
        queryLength={blastResultsMockData.query_len}
      />
    );
    await rendered;
  });

  it('should initially render overview', () => {
    const { asFragment, getByTestId } = rendered;
    expect(getByTestId('overview-hsp-detail')).toBeTruthy();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should load correct query and match sequence data', async () => {
    const { container } = rendered;
    const msa = container.querySelector('protvista-msa');
    expect(msa.data).toEqual([
      { name: 'Query', sequence: hsp.hsp_qseq },
      { name: 'Match', sequence: hsp.hsp_hseq },
    ]);
  });

  it('should change to wrapped and render when wrapped view is clicked', async () => {
    const { getByText, findByTestId, asFragment } = rendered;
    const wrappedButton = getByText('Wrapped');
    fireEvent.click(wrappedButton);
    expect(await findByTestId('wrapped-hsp-detail')).toBeTruthy();
    // skip the top level div, as it contains the dynamically injected style
    // that might be different across different runs ("sliding" effect)
    expect(asFragment().firstElementChild.firstElementChild).toMatchSnapshot();
  });
});
