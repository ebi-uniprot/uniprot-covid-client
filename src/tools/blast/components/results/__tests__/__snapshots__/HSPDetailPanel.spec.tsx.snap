// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HSPDetailPanel should change to wrapped and render when wrapped view is clicked 1`] = `
<div>
  <div
    class="hsp-detail-panel__header"
  >
    <h4>
      U6MKN0
    </h4>
    <button
      type="button"
    >
      <svg
        fill="currentColor"
        height="16"
        viewBox="0 0 352 512"
        width="16"
      >
        <path
          d="M243 256l100-100c12-12 12-32 0-45l-22-22a31 31 0 00-45 0L176 189 76 89a31 31 0 00-45 0L9 111a31 31 0 000 45l100 100L9 356a31 31 0 000 45l22 22c13 12 33 12 45 0l100-100 100 100c12 12 32 12 45 0l22-22c12-13 12-33 0-45L243 256z"
        />
      </svg>
    </button>
  </div>
  <div
    class="hsp-detail-panel__body"
  >
    <div
      class="button-group"
    >
      <div
        class="dropdown-container"
      >
        <button
          class="button dropdown tertiary"
          type="button"
        >
          Highlight properties
        </button>
        <div
          class="dropdown-menu"
        >
          <div
            class="dropdown-menu__content"
          >
            <ul>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Similarity
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Negative
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Positive
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Aliphatic
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Aromatic
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Charged
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Polar
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Serine Threonine
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Buried Index
                </button>
              </li>
              <li>
                <button
                  class="button primary"
                  type="button"
                >
                  Clustal
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Clustal 2
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Cinema
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Helix Propensity
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Hydro
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Lesk
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Mae
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Nucleotide
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Purine Pyrimidine
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Strand Propensity
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Taylor
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Turn Propensity
                </button>
              </li>
              <li>
                <button
                  class="button tertiary"
                  type="button"
                >
                  Zappo
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
      <fieldset
        class="msa-view-choice"
      >
        View:
        <label>
          <input
            checked=""
            name="view"
            type="radio"
          />
          Overview
        </label>
        <label>
          <input
            name="view"
            type="radio"
          />
          Wrapped
        </label>
      </fieldset>
    </div>
    <div>
      <div>
        <section
          class="hsp-detail-panel__visualisation hsp-detail-panel__visualisation__wrapped-row"
          data-testid="wrapped-hsp-detail"
        >
          <section
            class="hsp-label hsp-label--msa"
          >
            Alignment
          </section>
          <section
            class="hsp-detail-panel__visualisation--msa"
          >
            <protvista-msa
              colorscheme="clustal"
              height="40"
              length="45"
            />
          </section>
          <section
            class="hsp-label hsp-label--track"
          />
          <section
            class="hsp-detail-panel__visualisation--track"
          >
            <protvista-track />
          </section>
        </section>
      </div>
    </div>
  </div>
</div>
`;

exports[`HSPDetailPanel should initially render overview 1`] = `
<DocumentFragment>
  <div
    class="sliding-panel sliding-panel--bottom hsp-detail-panel"
    style="opacity: 0; margin-right: -1000px;"
  >
    <div>
      <div
        class="hsp-detail-panel__header"
      >
        <h4>
          U6MKN0
        </h4>
        <button
          type="button"
        >
          <svg
            fill="currentColor"
            height="16"
            viewBox="0 0 352 512"
            width="16"
          >
            <path
              d="M243 256l100-100c12-12 12-32 0-45l-22-22a31 31 0 00-45 0L176 189 76 89a31 31 0 00-45 0L9 111a31 31 0 000 45l100 100L9 356a31 31 0 000 45l22 22c13 12 33 12 45 0l100-100 100 100c12 12 32 12 45 0l22-22c12-13 12-33 0-45L243 256z"
            />
          </svg>
        </button>
      </div>
      <div
        class="hsp-detail-panel__body"
      >
        <div
          class="button-group"
        >
          <div
            class="dropdown-container"
          >
            <button
              class="button dropdown tertiary"
              type="button"
            >
              Highlight properties
            </button>
            <div
              class="dropdown-menu"
            >
              <div
                class="dropdown-menu__content"
              >
                <ul>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Similarity
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Negative
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Positive
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Aliphatic
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Aromatic
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Charged
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Polar
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Serine Threonine
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Buried Index
                    </button>
                  </li>
                  <li>
                    <button
                      class="button primary"
                      type="button"
                    >
                      Clustal
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Clustal 2
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Cinema
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Helix Propensity
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Hydro
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Lesk
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Mae
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Nucleotide
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Purine Pyrimidine
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Strand Propensity
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Taylor
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Turn Propensity
                    </button>
                  </li>
                  <li>
                    <button
                      class="button tertiary"
                      type="button"
                    >
                      Zappo
                    </button>
                  </li>
                </ul>
              </div>
            </div>
          </div>
          <fieldset
            class="msa-view-choice"
          >
            View:
            <label>
              <input
                checked=""
                name="view"
                type="radio"
              />
              Overview
            </label>
            <label>
              <input
                name="view"
                type="radio"
              />
              Wrapped
            </label>
          </fieldset>
        </div>
        <div>
          <section
            class="hsp-detail-panel__visualisation"
            data-testid="overview-hsp-detail"
          >
            <section
              class="hsp-label"
            >
              Overview
            </section>
            <div>
              <protvista-track
                height="9"
                highlight="418:446"
                layout="overlapping"
                length="486"
              />
              <protvista-track
                height="9"
                highlight="418:446"
                layout="overlapping"
                length="486"
              />
            </div>
            <section
              class="hsp-label"
            />
            <protvista-track
              highlight="418:446"
              layout="non-overlapping"
              length="486"
            />
            <section
              class="hsp-label"
            >
              Alignment
            </section>
            <protvista-manager
              attributes="displaystart displayend"
              displayend="28"
              displaystart="1"
            >
              <protvista-navigation
                length="28"
              />
              <protvista-msa
                colorscheme="clustal"
                height="40"
                length="28"
              />
            </protvista-manager>
          </section>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
